packages/loopring_v3/contracts/impl/agents/FastWithdrawalAgent.sol

```diff
@@ -0,0 +1,203 @@
+            keccak256(
+                abi.encodePacked(
+                    FASTWITHDRAWAL_TYPEHASH,
+                    fastWithdrawal.exchange,
+                    fastWithdrawal.from,
+                    fastWithdrawal.to,
+                    fastWithdrawal.token,
+                    fastWithdrawal.amount,
+                    fastWithdrawal.feeToken,
+                    fastWithdrawal.fee,
+                    fastWithdrawal.onchainFeePayment,
+                    fastWithdrawal.validUntil
+                )
+            )
+        );
+
+        // Make sure the tx isn't used multiple times
+        require(txHash[hash] == false, "TX_REPLAY");
+        txHash[hash] = true;
+
+        // Check the signature
+        require(hash.verifySignature(fastWithdrawal.from, fastWithdrawal.signature), "INVALID_SIGNATURE");
+
+        // Check the time limit
+        require(now <= fastWithdrawal.validUntil, "TX_EXPIRED");
+
+        // Transfer the tokens to the requested address
+        // using funds from `msg.sender`.
+        transfer(
+            fastWithdrawal.to,
+            fastWithdrawal.token,
+            fastWithdrawal.amount
+        );
+
+        // If the fee is paid offchain in the same token as the main transfer
+        // just add it to the transfer amount
+        if (!fastWithdrawal.onchainFeePayment && fastWithdrawal.token == fastWithdrawal.feeToken) {
+            fastWithdrawal.amount = fastWithdrawal.amount.add(fastWithdrawal.fee);
+            fastWithdrawal.fee = 0;
+        }
+
+        IExchangeV3 exchange = IExchangeV3(fastWithdrawal.exchange);
+
+        // Approve the offchain transfer from the withdrawing account back to the mediator (msg.sender)
+        exchange.approveConditionalTransfer(
+            fastWithdrawal.from,
+            msg.sender,
+            fastWithdrawal.token,
+            fastWithdrawal.amount
+        );
+
+        // Fee payment to the mediator (msg.sender)
+        if (fastWithdrawal.fee > 0) {
+            if (fastWithdrawal.onchainFeePayment) {
+                // Do fee payment directly from the user's wallet if requested
+                // (using the approval for the exchange)
+                exchange.onchainTransferFrom(
+                    fastWithdrawal.from,
+                    msg.sender,
+                    fastWithdrawal.feeToken,
+                    fastWithdrawal.fee
+                );
+            } else {
+                // Do the fee payment with an internal transfer as well
+                // Approve the fee transfer
+                exchange.approveConditionalTransfer(
+                    fastWithdrawal.from,
+                    msg.sender,
+                    fastWithdrawal.feeToken,
+                    fastWithdrawal.fee
+                );
+            }
+        }
+    }
+
+    function transfer(
+        address to,
+        address token,
+        uint    amount
+        )
+        internal
+    {
+        bool success;
```

Comment:
check `msg.sender != to`? I think we shoud skip transfers in this situation.

Reply:
Same reason as above. You pay a bit of gas to check these things for all fast withdrawals, but it will never be triggered because it's useless so you just make the implementation more complicated and very slightly less efficient for all valid transactions (which will be 100% of all fast withdrawals that actually get onchain).